### PR TITLE
Initial SQL .NET SDK

### PR DESCRIPTION
**Pull request by nathannfan**
## Description
Corresponding swagger checkin: https://github.com/Azure/azure-rest-api-specs/pull/622

This covers core SQL functionality, firewall rules, and replicationLinks.  

This is a revival of the pull request here: https://github.com/Azure/azure-sdk-for-net/pull/2493, but intended for release.

Version number is not correct - please suggest a proper version number.
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/dev/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/CONTRIBUTING.md#sdk-generation-guidelines)
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as a link to the swagger spec, used to generate the code.
- [x] The `project.json` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@nathannfan__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (nafan). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

**Comment by cormacpayne** [comment]
@nathannfan Hey Nathan, would you mind taking a look at the checklist provided above and filling out the corresponding boxes?

**Comment by nathannfan** [comment]
@cormacpayne Updated.  I was trying to do that earlier but for some reason I couldn't.  Also, I can't read the contributing guidelines.  Is it possible that it has moved?

**Comment by cormacpayne** [comment]
@nathannfan ah yes, good catch. We recently moved it from `.\CONTRIBUTING.md` to `.github\CONTRIBUTING.md`, which you can find [here](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md)

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/Database.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/Database.cs]: This list of max sizes should not be here. We should replace it with a suggestion to call the /capabilities API to get a list of max sizes per edition/slo/pool/dtu that are available.

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/Database.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/Database.cs]: remove list of possible values and reference to the /capabilities api

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/Database.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/Database.cs]: Same thing. Reference the /capabilities API to get available editions

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ElasticPool.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ElasticPool.cs]: ref /capabilities

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ElasticPool.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ElasticPool.cs]: ref /capabilities.  

All places please.

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ElasticPool.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ElasticPool.cs]: Nit: This should probably be "Azure SQL Database Elastic Pool" in all locations

Comment (akromm) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ElasticPool.cs]: find & replace "SQL Azure" -> "Azure SQL"

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/RecommendedIndex.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/RecommendedIndex.cs]: It would be good to reference RecommendedIndexActions class instead of writing out all the possible values here.

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/RecommendedIndex.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/RecommendedIndex.cs]: RecommendedIndexStates class.

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ServiceObjectiveName.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ServiceObjectiveName.cs]: This list is quite out of date. It would be good to remove it if possible.

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/TargetDatabaseEditions.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/TargetDatabaseEditions.cs]: Can references to this be replaced with DatabaseEditions and delete this class?

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/TargetElasticPoolEditions.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/TargetElasticPoolEditions.cs]: Same comment as for TargetDatabaseEditions

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Properties/AssemblyInfo.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Properties/AssemblyInfo.cs]: Lets change this to 1.0.0.0

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Properties/AssemblyInfo.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Properties/AssemblyInfo.cs]: "Microsoft Azure SQL Management Library"

File: src/ResourceManagement/SqlManagement/Sql.Tests/DatabaseCrudScenarioTests.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Sql.Tests/DatabaseCrudScenarioTests.cs]: Please add file header

File: src/ResourceManagement/SqlManagement/Sql.Tests/DatabaseReplicationLinkScenarioTests.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Sql.Tests/DatabaseReplicationLinkScenarioTests.cs]: Please add file header

File: src/ResourceManagement/SqlManagement/Sql.Tests/ElasticPoolActivityScenarioTests.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Sql.Tests/ElasticPoolActivityScenarioTests.cs]: Please add file header.  Please check all files in SQL.Tests.

**Review by akromm-zz — CHANGES_REQUESTED** [review]
As a note, don't forget to add the required files to generate the nuget package onces all the APIs are done.

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Properties/AssemblyInfo.cs
Comment (jaredmoo) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Properties/AssemblyInfo.cs]: What are the consequences of version `1.0.0.0`? Doesn't this mean we are committing to keeping API stable?

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/Database.cs
Comment (nathannfan) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/Database.cs]: This is generated from the swagger, so I will address this in a separate PR for the specs.

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/Database.cs
Comment (nathannfan) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/Database.cs]: This is generated from the swagger, so I will address this in a separate PR for the specs.

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/Database.cs
Comment (nathannfan) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/Database.cs]: This is generated from the swagger, so I will address this in a separate PR for the specs.

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ElasticPool.cs
Comment (nathannfan) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ElasticPool.cs]: This is generated from the swagger, so I will address this in a separate PR for the specs.

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ElasticPool.cs
Comment (nathannfan) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ElasticPool.cs]: This is generated from the swagger, so I will address this in a separate PR for the specs.

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ServiceObjectiveName.cs
Comment (nathannfan) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ServiceObjectiveName.cs]: The Azure SDK team had me add this.  I know it's hard to keep up-to-date, but that's why I have it as model-as-string, so values not in this list can also be specified.

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ElasticPool.cs
Comment (nathannfan) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ElasticPool.cs]: Addressed in Swagger.

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ElasticPool.cs
Comment (nathannfan) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ElasticPool.cs]: This is generated from the swagger, so I will address this in a separate PR for adding the capabilities api swagger.

**Comment by salameer** [comment]
@cormacpayne Captain Cormac Can we kindly review these PR :)

**Comment by shahabhijeet** [comment]
@salameer once we get a sign off from Adam or jared, we should be good to go.

File: src/ResourceManagement/SqlManagement/Sql.Tests/DatabaseCrudScenarioTests.cs
Comment (jaredmoo) [inline src/ResourceManagement/SqlManagement/Sql.Tests/DatabaseCrudScenarioTests.cs]: I know this is annoying to regenerate the replay data, but can you split each of these cases into separate [Fact]s?

File: src/ResourceManagement/SqlManagement/Sql.Tests/DatabaseCrudScenarioTests.cs
Comment (jaredmoo) [inline src/ResourceManagement/SqlManagement/Sql.Tests/DatabaseCrudScenarioTests.cs]: these can all be the same [Fact] since they are upgrading a single database through the test

**Review by jaredmoo — APPROVED** [review]

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ServiceObjectiveName.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/ServiceObjectiveName.cs]: I would recommend updating it to be current as of today then.

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Properties/AssemblyInfo.cs
Comment (akromm) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Properties/AssemblyInfo.cs]: We can still mark the NuGet package as "-prerelease" until all the swagger conversions are done.

The main motivation behind this is to make a distinction between hyak generated and swagger generated nuget packages.

**Review by akromm-zz — APPROVED** [review]

**Comment by nathannfan** [comment]
Thanks for the signoffs, guys.

Cormac, please don't merge this yet - I'd like to address all the minor issues and send one more commit before checking in.

**Comment by jaredmoo** [comment]
When I build, I get these warnings:

7>D:\git\azure-sdk-for-net\src\ResourceManagement\SqlManagement\Sql.Tests\project.json(26,43): warning NU1012: Dependency conflict. Microsoft.Rest.ClientRuntime.Azure.Authentication 2.2.10 expected Microsoft.Rest.ClientRuntime >= 2.3.5 < 3.0 but received 2.3.3
7>D:\git\azure-sdk-for-net\src\ResourceManagement\SqlManagement\Sql.Tests\project.json(26,43): warning NU1012: Dependency conflict. Microsoft.Rest.ClientRuntime.Azure.TestFramework 1.5.2 expected Microsoft.Rest.ClientRuntime >= 2.3.5 < 3.0.0 but received 2.3.3

File: src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/TargetDatabaseEditions.cs
Comment (nathannfan) [inline src/ResourceManagement/SqlManagement/Microsoft.Azure.Management.Sql/Generated/Models/TargetDatabaseEditions.cs]: Wow, this wasn't even referenced.

File: src/ResourceManagement/SqlManagement/Sql.Tests/DatabaseCrudScenarioTests.cs
Comment (nathannfan) [inline src/ResourceManagement/SqlManagement/Sql.Tests/DatabaseCrudScenarioTests.cs]: Discussed offline - because of the huge runtime and because these tests can be considered more E2E than unit tests, leaving like this for now.

**Comment by nathannfan** [comment]
@cormacpayne can this be merged in now?

**Pull request merged**
